_posts/2018-07-25-dota-macos-performance.md

```diff
@@ -0,0 +1,81 @@
+
+Note that technically, "Immediate" recording has less CPU overhead, as well as more explicit threading: no surprises at submit time. We consider it to be the superior mode, however the application needs to make sure to not hold onto the recorded command buffers for too long, and submit more often than once per frame. Otherwise, the latency kicks in, and Dota being programmed to allow at most a single frame of latency is able to produce less frames.
+
+Another interesting observation is that 71% of time spent by gfx/immediate (when recording the commands only) is actually inside the driver, according to the instrumented profile. This gives us an upper bound for Vulkan Portability overhead on Metal to be about 40%, although that includes some OS and Metal runtime bits as well.
+
+### Results
+
+| Test/Library             | gfx/immediate | gfx/deferred | MoltenVK | OpenGL      |
+| ------------------------ | ------------- | ------------ | -------- | ----------- |
+| CPU % of Main thread     | 36%           | 12%          | 21%      | ?           |
+| _dual-core platform_     | | | |
+| fps in the simple case   | 88 fps        | 105 fps      | 100 fps  | 105-110 fps |
+| fps on low settings      | 39.1 fps      | 45.7 fps     | 42.2 fps | 45.3 fps    |
+| fps variablility on low  | 5.0           | 5.6          | 5.7      | 5.5         |
+| fps on high settings     | 32.8 fps      | 38.7 fps     | 36.0 fps | 34.7 fps    |
+| fps variablility on high | 4.3           | 5.3          | 5.6      | 6.8         |
+| _quad-core platform_     | | | |
+| fps on low settings      | 63.1 fps      | 76.9 fps     | 68.3 fps | 79.6 fps    |
+| fps variablility on low  | 13.6          | 15.1         | 15.3     | 12.9        |
+| fps on high settings     | 54.6 fps      | 59.6 fps     | 65.6 fps | 51.9 fps    |
+| fps variablility on high | 11.9          | 11.6         | 10.9     | 6.1         |
+
+The first metric shows how much time of the main thread is spent inside the portability library, compared to the total execution time (which is the actual time minus all the sleeping). It was measured with Instruments on a simple scene ("fps in simple case" shows the frame rate on it). Note that the submission is done on a separate thread by Dota, so that time isn't taken into account here. Interestingly, the less time we spend on the main thread the faster our frame rate ends up being. Or, in other words, it's the race of who gets to the submission first :)
```

Comment:
Might be missing a word like "the", before the word "instruments". Also, likly [sic] shouldn't be capital.
>It was measured with **Instruments** on a simple scene ("fps in simple case" shows the frame rate on it).

Reply:
> Or, in other words, it's the race of who gets to the submission first :)

I think it should be:
> Or, in other words, it's a race of who gets to the submission first :)